src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadStack.cs

```diff
@@ -82,7 +82,12 @@ private void AddCurrent()
 
         public void Initialize(Type type, JsonSerializerOptions options, bool supportContinuation)
         {
-            JsonClassInfo jsonClassInfo = options.GetOrAddClass(type);
+            JsonClassInfo? jsonClassInfo = options.LastClass;
+            if (jsonClassInfo?.Type != type)
+            {
+                jsonClassInfo = options.GetOrAddClass(type);
+                options.LastClass = jsonClassInfo;
+            }
```

Comment:
Looks like this logic is duplicated? Can we extract into a helper?

Reply:
Yes moved to a helper. Thanks